vivarium/core/types.py

```diff
@@ -0,0 +1,9 @@
+from typing import Union, Tuple, NewType, Dict, Any
+
+
+Path = Union[Tuple[str, ...], Tuple[()]]
+Topology = NewType('Topology', dict)
+Schema = NewType('Schema', Dict[str, Any])
+State = NewType('State', Dict[str, Any])
+Update = NewType('Update', Dict[str, Any])
+CompositeDict = NewType('CompositeDict', Dict[str, Any])
```

Comment:
Yay for defining these! Named concepts will greatly assist people learning to use Vivarium. The type declarations also serve as very concise reference docs on how to build these values.

* Add `Config = Dict[str, Dict[str, Any]]`?
* Can we make these more specific than `dict` and `Dict[str, Any]`?
* Do we really want these to be `NewType` types or just type aliases? mypy treats a `NewType` like a distinct subclass, which means it auto-coerces to its base class but not vice versa. (The term "newtype" comes from Haskell where it creates a distinct new type that won't coerce to/from any other type. In various other languages, a struct type is distinct from another struct type even if it happens to contain the same element types. That's a "newtype.") So if `Schema` is a `NewType`, then all the code to construct Schemas has to call the "constructor" (a no-op function), e.g.: `Schema({'k1': 223, '_default': 0})` or `Schema(dict(k1=223, _default=0))` or something has to explicitly "convert" them.
* If it's fine to call constructors when building each of these values, then why not make them real `dict` subclasses (nominal types)? Then we could add methods like `__str__()`, debugging aids like optional runtime checks, places where you can put a breakpoint, etc.
  * Actually, Factory could auto-convert a `dict` to a `Schema`, `Config`, etc., allowing the caller to use the dict literal syntax while the rest of the code passes around actual `Schema` objects to aid debugging and type checking. In that case Factory would accept a `Union[Schema, Dict[str, Any]]` while most of the library code would accept `Schema`.
* There's a backward-compatible syntax for `TypedDict` so we can still use it to get mypy to type-check common dict items, e.g. that `_updater` holds a `Literal['accumulate', 'set', 'merge']`. A simpler, optional approach is to add utility functions, like say (rough idea):
  ```
  def subconfig(default: Any, updater: Literal['accumulate', 'set', 'merge'], **kwargs):
      return dict(kwargs, _default=default, _updater=updater)
  ```

Reply:
Re making types more specific: I think we're blocked by the fact that mypy [doesn't support recursive types](https://github.com/python/mypy/issues/731). Given that constraint, I don't see how to make them any more specific. I'm open to doing that though if you see a way!

Re NewType vs Class vs Alias: I agree with making these types into full classes eventually. Since mypy treats them as distinct types, we have to treat the type like a class anyway like this:

```python
def generate_topology(self):
    return Topology({})
```

Therefore it should be really easy to replace the `Topology` type with a full class later on if we want to. This is why I used NewType instead of just a type alias